pkg/provisioning/provision.go

```diff
@@ -0,0 +1,240 @@
+}
+
+func GetService(
+	ctx *context.Context,
+	logger log.CtxLogger,
+	plService *pipeline.Service,
+	connService *connector.Service,
+	procService *processor.Service,
+) *Service {
+	return &Service{
+		ctx:              ctx,
+		logger:           logger.WithComponent("provision"),
+		pipelineService:  plService,
+		connectorService: connService,
+		processorService: procService,
+	}
+}
+
+func (s *Service) ProvisionConfigFile(path string) error {
+	data, err := ioutil.ReadFile(path)
+	if err != nil {
+		return err
+	}
+
+	before, err := Parse(data)
+	if err != nil {
+		return err
+	}
+
+	got := EnrichPipelinesConfig(before)
+
+	for k, v := range got {
+		// skip non valid? multierror?
+		err = ValidatePipelinesConfig(v)
+		if err != nil {
+			return err
+		}
+		err = s.provisionPipeline(k, v)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (s *Service) provisionPipeline(id string, config PipelineConfig) error {
+	var destMap map[string]connector.DestinationState
+	var srcMap map[string]connector.SourceState
+	var err error
+	oldPl, err := s.pipelineService.Get(*s.ctx, id)
+	if err == nil {
+		update = true
+		destMap, srcMap, err = s.copyStatesAndDeletePipeline(oldPl, config)
+		if err != nil {
+			return err
+		}
+	} else {
+		update = false
+	}
+
+	newPl, err := s.createPipeline(id, config)
+	if err != nil {
+		return err
+	}
+	for k, cfg := range config.Connectors {
+		err := s.createConnector(id, k, cfg, destMap, srcMap)
+		if err != nil {
+			return err
+		}
+		_, err = s.pipelineService.AddConnector(*s.ctx, newPl, k)
+		if err != nil {
+			return err
+		}
+		for k1, cfg1 := range cfg.Processors {
+			err := s.createProcessor(id, processor.ParentTypeConnector, k1, cfg1)
+			if err != nil {
+				return err
+			}
+			_, err = s.connectorService.AddProcessor(*s.ctx, k, k1)
+			if err != nil {
+				return err
+			}
+		}
+	}
+	for k, cfg := range config.Processors {
+		err := s.createProcessor(id, processor.ParentTypePipeline, k, cfg)
+		if err != nil {
+			return err
+		}
+		_, err = s.pipelineService.AddProcessor(*s.ctx, newPl, k)
+		if err != nil {
+			return err
+		}
+	}
+
+	// set pipeline status
+	status := pipeline.StatusUserStopped
+	if config.Status == StatusRunning {
+		err := s.pipelineService.Start(*s.ctx, s.connectorService, s.processorService, newPl)
+		if err != nil {
+			return err
+		}
+		status = pipeline.StatusRunning
+	}
+	newPl.Status = status
+
+	return nil
+}
+
+func (s *Service) copyStatesAndDeletePipeline(pl *pipeline.Instance, config PipelineConfig) (map[string]connector.DestinationState, map[string]connector.SourceState, error) {
+	destMap := make(map[string]connector.DestinationState, 1)
+	srcMap := make(map[string]connector.SourceState, 1)
+
+	for k := range config.Connectors {
+		destState, srcState, err := s.getConnectorState(k)
+		if err != nil {
+			return nil, nil, err
+		}
+		destMap[k] = destState
+		srcMap[k] = srcState
+	}
+
+	err := s.pipelineService.Delete(*s.ctx, pl)
```

Comment:
This won't be so easy, we need to first remove all processors (on the pipeline and connectors) and all connectors.